062-quotas-management.md

```diff
@@ -0,0 +1,167 @@
+   The only way to configure them right now is imperatively use the Kafka Admin API.
+   While it can be used by users, it does not follow Strimzi's basic idea of declarative management.
+   Having a dedicated API for quota configuration as part of the `Kafka` CR would make it much easier to configure the default quotas.
+2. The initial releases of the Strimzi Quota Plugin were simple, preventing users from producing messages directly to a broker running out of disk space. 
+   However, these versions did not address the issue of replicated messages produced on other brokers in the Kafka cluster, which could still fill up the disk space of the destination broker."
+   They also didn't support many useful features, such as JBOD storage.
+   Because the initial releases were so simple, it was possible to configure them in the `.spec.kafka.config` section of the `Kafka` CR.
+   The next - not yet released - version of the Strimzi Quota plugin brings many improvements.
+   More details about how the new version works can be found in the [Cluster Wide Volume Usage Quota Management](./047-cluster-wide-volume-usage-quota-management.md) proposal.
+   It operates at the level of a whole Kafka cluster and it has improved capability in preventing replicated messages from filling in broker storage.
+   It also supports JBOD storage.
+   But it also has a much more complicated configuration.
+   The plugin uses the Kafka Admin API to gather information about disk capacity from the Kafka brokers.
+   So, it needs to connect to the Kafka cluster as a client.
+   Having the user configure this directly in the `.spec.kafka.config` section would add a lot more complexity, as the user would need to configure the credentials for securing a connection,  such as keystores, truststores, passwords, bootstrap addresses, and so on.
+   Having a dedicated API for quota configuration would make it much easier to have the operator configure the client automatically.
+
+## Proposal
+
+This proposal suggests adding a new `quotas` API to the `Kafka` custom resource.
+It will be inside the `.spec.kafka` section.
+And it will allow users to configure different types of quotas.
+As with existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The built-in Kafka plugin
+2. The Strimzi Quota plugin
+
+The first supported plugin - Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow you to configure the default user quotas for this plugin in the same format as the `KafkaUser` API currently supports:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
```

Comment:
I was wondering if an opaque list isn't better than specific fields (I was thinking at the tiered storage proposal doing this way).

Reply:
It is just 4 fields -> We have it defined like this also in the `KafkaUser` CRD, it gives us better validation etc. I do not have a strong opinion if more people want it as a Map, but I would probably prefer to keep the dedicated fields.

Reply:
Well they are just for filed for this type of plugin. They could be more for another type. Maybe I am overthinking?

Reply:
I don't know. We use the non-validated `<String, String>` maps normally in places where we cannot really use proper fields:
* Because there are too many options to model each one in the API (e.g. `.spec.kafka.config`)
* Because the list of options is not finite and new options can be part of plugins or new releases (e.g. `.spec.kafka.config`)
* Because we add support for a custom plugin and thus do not know the configuration options (this would be the tiered storage for example)

We do not use it in places like this where we have few clearly defined options.

You are right that there might be another quota plugin somewhere in the future. But given how this proposal goes, I would not see it happen anytime soon 😉. Also, such a plugin would have its own type and configurations anyway.

Reply:
Ok let's move on with this for now.